### PR TITLE
[#1599] Forked server.concurrent from Xtext

### LINKED ISSUE #1599: Fork xtext.ide.server.concurrent
The request manager should become a aware of dependencies between requests. Therefore it needs to be adjusted. We'll start with a 1:1 for of the Xtext impl, but migrated to Java.

### PR DESCRIPTION
**Pull request by szarnekow**
closes #1599

### CODE REVIEW

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XLanguageServerImpl.java [hunk 1]
```diff
@@ -243,8 +244,12 @@ public void setWorkspaceManager(XWorkspaceManager manager) {
 
 		access.addBuildListener(this);
 
+		// Stopwatch sw = Stopwatch.createStarted();
 		workspaceManager.initialize(baseDir);
+		// System.out.println("workspaceManager.initialize " + sw);
+		// sw = Stopwatch.createStarted();
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XLanguageServerImpl.java]: Commented lines are still necessary?

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSRequestManager.java [hunk 2]
```diff
@@ -13,16 +13,16 @@
...
-import org.eclipse.xtext.ide.server.concurrent.AbstractRequest;
-import org.eclipse.xtext.ide.server.concurrent.RequestManager;
+import org.eclipse.n4js.ide.xtext.server.concurrent.XAbstractRequest;
+import org.eclipse.n4js.ide.xtext.server.concurrent.XRequestManager;
 
 /**
  * Prints stack traces for requests that threw an exception
  */
-public class N4JSRequestManager extends RequestManager {
+public class N4JSRequestManager extends XRequestManager {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSRequestManager.java]: Could you integrate this class into `XRequestManager`?

Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/N4JSRequestManager.java]: 👍

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XLanguageServerImpl.java (hunk 1 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XLanguageServerImpl.java]: :+1:

**Comment by szarnekow** [comment]
Please take another look at this PR @mmews-n4

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XWorkspaceManager.java [hunk 3]
```diff
@@ -313,6 +314,12 @@ public XProjectManager getProjectManager(URI uri) {
 		String name = null;
 		if (projectConfig != null) {
 			name = projectConfig.getName();
+		} else if (N4Scheme.isN4Scheme(uri)) {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XWorkspaceManager.java]: The intention was to keep X-classes free of N4JS related logic.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSRuntimeModule.java [hunk 4]
```diff
@@ -204,6 +206,14 @@
 		return ErrorAwareLinkingService.class;
 	}
 
+	/**
+	 * Bind the {@link ILocationInFileProvider} that is aware of derived elements.
+	 */
+	@Override
+	public Class<? extends ILocationInFileProvider> bindILocationInFileProvider() {
+		return N4JSLocationInFileProvider.class;
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSRuntimeModule.java]: This is for navigation to the correct line? Why isn't it needed in the ui case anymore?

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XWorkspaceManager.java (hunk 3 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XWorkspaceManager.java]: Ok, I'll extract a subclass in a subsequent step

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSRuntimeModule.java (hunk 4 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSRuntimeModule.java]: The UI injector "inherits" it from the runtime injector

Comment (szarnekow) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSRuntimeModule.java]: Should have been bound here right from the start

Comment (mmews-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/N4JSRuntimeModule.java]: I see.

**Pull request merged**